# Tutorial objectives

The **Summary statistics and QC tutorial** is intended as a functional guide to help assess the quality characteristics of a single Nanopore sequencing run. This tutorial aims to enable an objective assessment of the performance of a Nanopore flowcell run and to assess the sequence characteristics to benchmark quality. 

Sufficient information is provided in the tutorial that the workflow can be tested, validated and replicated. The tutorial is provided with an example dataset from a barcoded sequence library. The tutorial is intended to address important questions;

* how many reads (and how many gigabases) were sequenced?
* what fraction of my sequence collection is good quality?
* how are longer sequence reads represented in my sample?
* how uniform is the representation of different barcodes?

**Methods used** in this tutorial include 

* **`R`** for statistical analysis and reporting
* **`sequencing_summary.txt`** as data source for parsing

**Computational requirements** for this tutorial include 

* Computer running Linux (Centos7, Ubuntu 18_10, Fedora 29) 
* At least 8 Gb RAM 
* Runtime with provided example data - approximately 10 minutes


\pagebreak

# Quick start 

1. Most software dependecies are managed though **`conda`**, install as described at  <br> [https://conda.io/docs/install/quick.html](https://conda.io/docs/install/quick.html). Accept the license agreement during installation, and it is recommended to allow the Conda installer to prepend its path to your `.bashrc` file when asked.
```
    wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
    bash Miniconda3-latest-Linux-x86_64.sh
    bash
```
2. Download Nanopore QC tutorial & example files into a folder named `QCTutorial`. This tutorial requires the **`git-lfs`** large file support capabilities; this should be installed first through **`conda`**
```
    conda install -c conda-forge git-lfs
    git lfs install
    git clone https://github.com/nanoporetech/ont_tutorial_basicqc.git QCTutorial
```
3. Change working directory into the new `QCTutorial` folder
```
    cd QCTutorial
```
4. Install conda software dependencies into a discrete conda environment
```
    conda env create --name BasicQC --file environment.yaml
```
5. Initialise the new conda environment with 
```
    source activate BasicQC
```
6. *Optional:* edit the provided **`config.yaml`** file to match your own study design
7. Render the report using results from the analysis above
```
    R --slave -e 'rmarkdown::render("Nanopore_SumStatQC_Tutorial.Rmd", "html_document")'
```

This workflow will create an HTML file in the working directory called **`Nanopore_SumStatQC_Tutorial.html`**. Open this file in your prefered web browser to view the summary statistics and a graphical review of your sequence collection.


\pagebreak




# Introduction

This tutorial aims to summarise the data characteristics from an Oxford Nanopore Technologies sequencing run. Observations from basecalled reads and their quality characteristics, temporal performance, and barcoded content are presented. The information presented is derived solely from the **`sequence_summary.txt`** file produced during basecalling with the Guppy software. A compatible file will also be generated by MinKNOW in upcoming releases.

This tutorial document has been produced from an **`Rmarkdown`** template. This template is intended as the starting point for your exploration and assessement of Oxford Nanopore DNA sequence data. The goals from this tutorial include

1. To introduce a literate framework for analysing base-calling summary statistics to evaluate the relative performance of runs
1. To provide basic QC metrics so that a review and consideration of experimental data can be undertaken 
1. To provide training as to which QC metrics are of most interest and to encourage an understanding of how different aspects of sequence data quality can be attributed to sample characteristics from DNA isolation and library preparation.

Several of the plots included in this report have been replicated from publicly available projects such as POREquality [^1], minion_qc [^2], and pycoQC [^3]. 

[^1]: [jcarsweshau/POREquality](https://github.com/carsweshau/POREquality) 
[^2]: [roblanf/minion_qc](https://github.com/roblanf/minion_qc)
[^3]: [a-slide/pycoQC](https://github.com/a-slide/pycoQC)

The **`sequence_summary.txt`** file is automatically produced during base-calling with the Guppy software. This summary file contains rich metadata for each sequence read produced during a run. These data include timestamp, pore duration, read quality, and channel information, in addition to the characteristics of the resulting DNA sequence. This tutorial uses this summary file for performance reasons. 

Tools such as wub [^4] utilise the **`fastq`** files for quality metrics, and other tools make extensive use of the **`fast5`** files. Parsing the **`fast5`** files provides additional analytical context but is much more demanding in terms of compute resource and time. This tutorial is lightweight and is intended to run within a few minutes on a desktop computer.

[^4]: [nanoporetech/wub](https://github.com/nanoporetech/wub)


# Setup a computational environment for your QC analysis

This tutorial is intended to be simple to install, run and customise. The analysis is performed using the **`R`** statistical software and further functionality is provided by a number of **`R packages`** and the **`RStudio`** graphical user interface. The following steps describe a simple approach to installing the tutorial and its dependencies.

## Conda package management software

**`Conda`** provides simple software package management. A wide variety of bioinformatics and scientific computing software has been deployed within Conda and it provides a streamlined way to install both software packages and their required dependencies without the requirement for administrative rights. These installation instructions assume that you are using the **`BASH`** shell interface to your computer.

Install the Conda software on your computer using the instructions provided at [https://conda.io/docs/install/quick.html](https://conda.io/docs/install/quick.html)

A recommended **`Conda`** installation could be installed on Linux with the following commands

```
# download Python3 version of the Miniconda installer
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh

bash Miniconda3-latest-Linux-x86_64.sh

bash
```

The commands here are used to (1) download the Conda installer, (2) Install conda and (3) reload your bash shell, and make the new Conda installation available.

We recommend that the Conda installer be allowed to prepend the Conda path to your **`.bashrc`** file. This makes Conda available each time  you log in to the system. If Conda does not add this path, you should either edit your own `.bashrc` file or initialise your Conda environment manually with the instructions provided.

Check that Conda has been installed with the commands

```
echo $PATH
conda --version
```


## Download the tutorial files

The tutorial documents and example data are contained on the [Oxford Nanopore Technologies Github site](www.github.com/nanoporetech). This Basic QC tutorial is contained within a project called **`ont_tutorial_basicqc`**. The tutorial download requires the **`large file support (lfs)`** extensions to **`git`** - the summary files from **`Guppy`** base-calling and de-multiplexing that are distributed with the tutorial are relatively large files.

The installation of the **`git-lfs`** extensions and download of the tutorial files and accompanying dataset can be performed with the commands

```
conda install -c conda-forge git-lfs
git lfs install
git clone https://github.com/nanoporetech/ont_tutorial_basicqc.git QCTutorial

# Change into the created `QCTutorial` directory
cd QCTutorial

```

This will download a collection of files into a new folder called `QCTutorial`. The files downloaded include

1. The Rmarkdown file, **`Nanopore_SumStatQC_Tutorial.Rmd`**, includes this documentation and performs the analysis.
1. **`RawData/lambda_sequencing_summary.txt.bz2`** is a summary statistic file describing approximately 1 million sequence reads from a barcoded lambda DNA sequencing run. These sequence reads have been base-called using the Guppy software. This file has been bzip2 compressed.
1. **`RawData/lambda_barcoding_summary.txt.bz2`** is a barcode summary statistic file describing the barcode assignments for the sequence reads. This barcode assignment has been performed using the Guppy software.
1. **`Static/`** is a folder containing some of the descriptive images used in the tutorial, text for the tutorial, bibliography and style-sheets.
1. **`environment.yaml`** is a text file (in yaml format) that describes the software packages and computational environment that will be used to build a working compute environment using Conda.
1. **`config.yaml`** is a text file (in yaml format) that describes the sequence summary statistic files to be analysed. This is the main file to be edited to describe your own analysis.


## Build a Conda environment

In the previous section we downloaded the project files. This download includes the file, `environment.yaml` that can be used to initialise a **`Conda`** working environment. To create a Conda environment arbitrarily named `BasicQC` we should use the commands

```
conda env create --name BasicQC --file environment.yaml
source activate BasicQC
```


# Run the analysis


Analysis of the sequences specified within the **`Rmarkdown`** file will be performed as part of the **`knit`** process. This will load the **`summary_statistics`** file, will prepare a sequence analysis, render figures and prepare the report. To start the analysis, it is only necessary to click the  **`knit`** button in the **`Rstudio`** software - please see figure \ref{fig:KnitIt} below. The **`Rstudio`** software can be opened with the tutorial markdown document with the command 

```
rstudio Nanopore_SumStatQC_Tutorial.Rmd
```

![](Static/Images/KnitIt.png)


It is possible to perform the QC analysis using **`knit`** from the command line; the commands below will knit the template document to produce and html format report with the command.

```
R --slave -e 'rmarkdown::render("Nanopore_SumStatQC_Tutorial.Rmd", "html_document")'
```

This command will build the HTML document that contains the analysis and exploration of the sequence collection. The file **`Nanopore_SumStatQC_Tutorial.html`** should be opened in a web-browser.

\pagebreak
